File: src/ipc/items.ts
/**
 * `core_*` items Tauri コマンドの型付きラッパー
 * (`src-tauri/src/commands/items.rs`)。
 *
 * `module_id` を引数で渡し、バックエンド側で `state.modules` から `ModuleBackend` を
 * 引き当てる。フロントは ID 文字列だけ持って各 IPC を呼ぶ設計
 * (`module-contract.md` §6.2: `core_*` は Shell / 共通フックからのみ呼ぶ)。
 */
import { invoke } from "@tauri-apps/api/core";

import type { Item, ModuleId } from "@/lib/types";

export function listItemSummaries(input: {
  moduleId: ModuleId;
  projectId: string;
  limit?: number;
  offset?: number;
}) {
  return invoke<import("@/lib/types").ItemSummary[]>("core_list_item_summaries", {
    ...input,
    limit: input.limit ?? 100,
    offset: input.offset ?? 0,
  });
}

export function listItems(input: {
  moduleId: ModuleId;
  projectId: string;
  limit?: number;
  offset?: number;
}): Promise<Item[]> {
  return invoke<Item[]>("core_list_items", {
    moduleId: input.moduleId,
    projectId: input.projectId,
    limit: input.limit ?? 100,
    offset: input.offset ?? 0,
  });
}

/**
 * 並び替えなどスコープ内の全 ID が必要な画面向けに、100 件ずつ全ページを取得する。
 */
export async function listAllItems(input: {
  moduleId: ModuleId;
  projectId: string;
}): Promise<Item[]> {
  const pageSize = 100;
  const items: Item[] = [];
  let offset = 0;

  while (true) {
    const page = await listItems({ ...input, limit: pageSize, offset });
    items.push(...page);
    if (page.length < pageSize) return items;
    offset += page.length;
  }
}

/**
 * payload を含まない一覧 (`ItemSummary`) を 100 件ずつ全ページ取得する。大きな payload を持つ
 * モジュール (Mermaid / Diagram / Vector) の文書選択など、表示に payload が要らない画面向け。
 * 並び順はバックエンドの `updated_at DESC, id DESC`。
 */
export async function listAllItemSummaries(input: {
  moduleId: ModuleId;
  projectId: string;
}): Promise<import("@/lib/types").ItemSummary[]> {
  const pageSize = 100;
  const items: import("@/lib/types").ItemSummary[] = [];
  let offset = 0;

  while (true) {
    const page = await listItemSummaries({ ...input, limit: pageSize, offset });
    items.push(...page);
    if (page.length < pageSize) return items;
    offset += page.length;
  }
}

export function getItem(input: { moduleId: ModuleId; itemId: string }): Promise<Item> {
  return invoke<Item>("core_get_item", {
    moduleId: input.moduleId,
    itemId: input.itemId,
  });
}

export function createItem(input: {
  moduleId: ModuleId;
  projectId: string;
  title: string;
  tags: string[];
  payload: unknown;
}): Promise<string> {
  return invoke<string>("core_create_item", {
    moduleId: input.moduleId,
    projectId: input.projectId,
    title: input.title,
    tags: input.tags,
    payload: input.payload,
  });
}

export function updateItem(input: {
  moduleId: ModuleId;
  itemId: string;
  title: string;
  tags: string[];
  payload: unknown;
}): Promise<void> {
  return invoke<void>("core_update_item", {
    moduleId: input.moduleId,
    itemId: input.itemId,
    title: input.title,
    tags: input.tags,
    payload: input.payload,
  });
}

export function deleteItem(input: { moduleId: ModuleId; itemId: string }): Promise<void> {
  return invoke<void>("core_delete_item", {
    moduleId: input.moduleId,
    itemId: input.itemId,
  });
}

/**
 * `(projectId, moduleId)` スコープ内の items を `orderedIds` の順序で並び替える
 * (`docs/data-model.md` §6.5、PR-Y / ADR-0011)。
 *
 * 楽観的更新パターン: 呼び出し前にローカル state を `arrayMove` で更新済とし、
 * 失敗時は refetch (list 再取得) で旧順序に巻き戻す (Sidebar D&D と同パターン、
 * `docs/ui-design.md` §3.3.1)。
 */
export function reorderItems(input: {
  projectId: string;
  moduleId: ModuleId;
  orderedIds: string[];
}): Promise<void> {
  return invoke<void>("core_reorder_items", {
    projectId: input.projectId,
    moduleId: input.moduleId,
    orderedIds: input.orderedIds,
  });
}

File: src-tauri/src/commands/items.rs
//! items CRUD の Tauri コマンド (`core_*`、`module-contract.md` §6.2)。
//!
//! `module_id` をフロントから受け取って `state.modules` から `Arc<dyn ModuleBackend>` を引き、
//! `state.storage.scoped_for(module)` で `ScopedStorage` を生成して呼ぶ。
//!
//! `core_get_item` は Eager-on-Read (ADR-0006) を発火する。`core_list_items` は
//! 発火しない (`data-model.md` §7.2)。

use serde_json::Value as JsonValue;
use tauri::State;

use crate::error::AppError;
use crate::state::AppState;
use crate::storage::types::{Item, ItemId, ProjectId};

/// プロジェクト内の `module_id` 配下 items を `updated_at DESC, id DESC` 順で取得
/// (`StorageService::list_items`)。Eager-on-Read は発火させない。
#[tauri::command]
pub fn core_list_items(
    state: State<'_, AppState>,
    module_id: String,
    project_id: String,
    limit: u32,
    offset: u32,
) -> Result<Vec<Item>, AppError> {
    let module = state
        .module(&module_id)
        .ok_or_else(|| AppError::ModuleNotFound {
            module_id: module_id.clone(),
        })?;
    let scoped = state.storage.clone().scoped_for(module);
    scoped.list_items(&ProjectId::new(project_id), limit, offset)
}

/// item を 1 件取得 (Eager-on-Read 経由、ADR-0006 / `data-model.md` §7.2)。
#[tauri::command]
pub fn core_get_item(
    state: State<'_, AppState>,
    module_id: String,
    item_id: String,
) -> Result<Item, AppError> {
    let module = state
        .module(&module_id)
        .ok_or_else(|| AppError::ModuleNotFound {
            module_id: module_id.clone(),
        })?;
    let scoped = state.storage.clone().scoped_for(module);
    scoped.get_item(&ItemId::new(item_id))
}

/// 新規 item を作成。`payload` はモジュール固有の JSON でフロントから送られる。
/// `validate_payload` / `index_text` / `current_payload_version` はモジュール側が決める
/// (`ScopedStorage::create_item` 内で実行)。
#[tauri::command]
pub fn core_create_item(
    state: State<'_, AppState>,
    module_id: String,
    project_id: String,
    title: String,
    tags: Vec<String>,
    payload: JsonValue,
) -> Result<ItemId, AppError> {
    let module = state
        .module(&module_id)
        .ok_or_else(|| AppError::ModuleNotFound {
            module_id: module_id.clone(),
        })?;
    let scoped = state.storage.clone().scoped_for(module);
    scoped.create_item(&ProjectId::new(project_id), &title, &tags, payload)
}

/// item を更新 (ユーザー編集)。`data_revision` を **+1**。
#[tauri::command]
pub fn core_update_item(
    state: State<'_, AppState>,
    module_id: String,
    item_id: String,
    title: String,
    tags: Vec<String>,
    payload: JsonValue,
) -> Result<(), AppError> {
    let module = state
        .module(&module_id)
        .ok_or_else(|| AppError::ModuleNotFound {
            module_id: module_id.clone(),
        })?;
    let scoped = state.storage.clone().scoped_for(module);
    scoped.update_item(&ItemId::new(item_id), &title, &tags, payload)
}

/// item を物理削除。
#[tauri::command]
pub fn core_delete_item(
    state: State<'_, AppState>,
    module_id: String,
    item_id: String,
) -> Result<(), AppError> {
    let module = state
        .module(&module_id)
        .ok_or_else(|| AppError::ModuleNotFound {
            module_id: module_id.clone(),
        })?;
    let scoped = state.storage.clone().scoped_for(module);
    scoped.delete_item(&ItemId::new(item_id))
}

/// `(project_id, module_id)` スコープ内の items を `ordered_ids` の順序で並び替える
/// (`StorageService::reorder_items`、PR-Y / `data-model.md` §6.5)。
///
/// - `ordered_ids` は当該スコープの **全 item ID が過不足なく** 含まれる必要がある
/// - 1 トランザクションで全件 UPDATE → `data_revision +1`
/// - `updated_at` は不変
/// - `module_id` の存在は ModuleRegistry でチェック (未登録 module は `AppError::ModuleNotFound`)
#[tauri::command]
pub fn core_reorder_items(
    state: State<'_, AppState>,
    project_id: String,
    module_id: String,
    ordered_ids: Vec<String>,
) -> Result<(), AppError> {
    // module の存在チェック (登録されていない module への reorder は弾く)
    if state.module(&module_id).is_none() {
        return Err(AppError::ModuleNotFound {
            module_id: module_id.clone(),
        });
    }
    let pid = ProjectId::new(project_id);
    let ids: Vec<ItemId> = ordered_ids.into_iter().map(ItemId::new).collect();
    state.storage.reorder_items(&pid, &module_id, &ids)
}
